### ISSUE TITLE
Strange behavior of library files in runtimepath

### ISSUE THREAD

**Issue by dimchee** [noib3/nvim-oxi#86]
After following [crate setup section in examples](https://github.com/noib3/nvim-oxi/tree/master/examples#crate-setup), everything worked, until I tried standard plugin refresh without neovim restart. The problem can be simulated with next lua snippet:
```lua
-- crash.lua
vim.o.rtp = '.,' .. vim.o.rtp
package.loaded.foo = nil
Foo = require'foo'
package.loaded.foo = nil
Foo = require'foo'
```
Which on execution crashes, and reports: `Illegal instruction (core dumped)`. I tried different neovim versions - every version crashes. I suspect neovim has problem with loading library files (which can be further tested with compiled C library). Another thing I found out is that runtimepath has strange behavior when shared object files are concerned (Order of finding plugins is not respected). I am posting this Issue here because I think it affects usage of this plugin more than neovim itself. Maybe I can dig deeper (I am not very experienced but would like to help).

**Comment by AregevDev** [comment]
I am having the same issue. I am using the crate setup example on Windows and Neovim crashes when loading the DLL file.

**Comment by ricardokl** [comment]
My problem could be related to this.

 I tried reproducing a colorscheme plugin using nvim-oxi. I can load the library and apply the colorscheme, I can even use a "toggle" to change the colorscheme. I added a lua file at the "colors" folder, and I can call it using the "colorscheme" command.

The problem happens when I call a different colorscheme and try to call this one back again. This crashes Neovim.

I tried with and without setting "package.loaded" to nil. I tried having Rust manage a global config, tried handing it over to Neovim via "api::set_var", and tried not storing any config but just calling the default everywhere. Neovim still crashes.

I'm not sure what else I can try, but reading this here makes me think it is really something with how Neovim handles libs. Or maybe something regarding cleanup in the Neovim/nvim-oxi interaction?

I'm using the main branch here, with Neovim 0.10.4

**Comment by ricardokl** [comment]
Ok, right after my post I got a helpfull suggestion from an LLM looking into "entrypoint.rs". The issue is requiring the lib everytime. Instead, at initialization I store  "loaded" state in a global variable, and checking this I can decide not to load it again. It looks like this:

```lua
if not vim.g.my_plugin then vim.g.my_plugin = require("my_plugin") end
vim.g.my_plugin.load()
```

This seems to solve the crash.

The LLM seems to think it has to do with "arena_init()", which indeed panics if called more then once. Maybe there is a different way of handling this?

**Comment by ricardokl** [comment]
A possible change is the following in the arena init:

```rust
pub fn arena_init() {
    ARENA.with(|arena| {
        if arena.get().is_some() {
            // Arena is already initialized
            return;
        } else {
            let _ = arena.set(Arena::new());
        }
    });
}
```

This replaces the panic with a return, so that nothing is done if the arena is already initialized. Something similar can be done for the luajit and libuv init.

I'm not sure this creates any unwanted behavior, but there was no apparent problem in my test.

**Comment by noib3** [comment]
> A possible change is the following in the arena init:

`let _ = arena.set(Arena::new());` should be enough, even without the if (which doesn't even guarantee anything due to a possible TOCTOU race condition).

I'd be happy to merge a PR for this (w/ the equivalent for `luajit::init()` and `libuv::init()`.

**Comment by ricardokl** [comment]
Yes, this is much simpler. One final suggestion and I'll do this way: what about

```rust
pub fn arena_init() {
    ARENA.with(|arena| {
        if arena.get().is_none() {
             let _ = arena.set(Arena::new());
        }
    });
}
```

Only difference is this is a simple check instead of letting it fail.

Perhaps adding back a panic in case something unforseen happens?

**Comment by noib3** [comment]
What's the purpose of the `if`? Why not just:

```rust
pub fn arena_init() {
    ARENA.with(|arena| {
        let _ = arena.set(Arena::new());
    });
}
```

> would it?

Yes.

**Comment by ricardokl** [comment]
The "if" was for trying to have nothing happening in case there was something there already and I thought having only a single branch could allow this. The difference from the previous code is just that the init is on the first (and only) "if" branch.

The solution without "if" checks is the only way to avoid TOCTOU then, ok.
